cupyx/scipy/spatial/delaunay_2d/_tri.py

```diff
@@ -0,0 +1,570 @@
+            (self.max_triangles, 2), -1, dtype=cupy.int32)
+        self._tri_msg_sz = 0
+        self._org_flip_num = []
+
+    def _dispatch_check_delaunay(self, check_mode, org_act_num, tri_vote):
+        if check_mode == CheckDelaunayMode.CircleFastOrientFast:
+            check_delaunay_fast(
+                self._act_tri, self.triangles, self.triangle_opp,
+                self.triangle_info, tri_vote, org_act_num, self.n_points - 1,
+                self.point_vec, self.pred_consts)
+        elif check_mode == CheckDelaunayMode.CircleExactOrientSoS:
+            exact_check_vi = self.tri_msg
+            self._renew_counters()
+
+            check_delaunay_exact_fast(
+                self._act_tri, self.triangles, self.triangle_opp,
+                self.triangle_info, tri_vote, exact_check_vi, org_act_num,
+                self.counters, self.n_points - 1, self.point_vec,
+                self.pred_consts)
+
+            check_delaunay_exact_exact(
+                self.triangles, self.triangle_opp, tri_vote, exact_check_vi,
+                self.counters, self.n_points - 1, self.point_vec,
+                self.points_idx, self.pred_consts)
+
+    def _do_flipping(self, check_mode):
+        tri_num = self.triangles.shape[0]
+        if self._act_tri_mode == ActTriMode.ActTriMarkCompact:
+            active_tri = is_tri_active(self.triangle_info)
+            self._act_tri = cupy.arange(0, active_tri.shape[0])
+            self._act_tri = self._act_tri[active_tri].astype(cupy.int32)
+        elif self._act_tri_mode == ActTriMode.ActTriCollectCompact:
+            self._act_tri = self._act_tri[self._act_tri >= 0]
+
+        org_act_num = self._act_tri.shape[0]
+
+        # Check actNum, switch mode or quit if necessary
+        if org_act_num == 0:
+            # No more work
+            return False
+
+        if (check_mode != CheckDelaunayMode.CircleExactOrientSoS and
+                org_act_num < 64 * 32):
+            # Little work, leave it for the Exact iterations
+            return False
+
+        # See if there's little work enough to switch to collect mode.
+        if (org_act_num < 512 * 128 and
+                org_act_num * 2 < self.max_triangles and
+                org_act_num * 2 < tri_num):
+            self._act_tri_mode = ActTriMode.ActTriCollectCompact
+        else:
+            self._act_tri_mode = ActTriMode.ActTriMarkCompact
+
+        # Vote for flips
+        tri_vote = cupy.full(tri_num, 0x7FFFFFFF, dtype=cupy.int32)
+        self._dispatch_check_delaunay(check_mode, org_act_num, tri_vote)
+
+        # Mark rejected flips
+        flip_to_tri = cupy.empty(org_act_num, dtype=cupy.int32)
+        mark_rejected_flips(
+            self._act_tri, self.triangle_opp, tri_vote, self.triangle_info,
+            flip_to_tri, org_act_num)
+
+        del tri_vote
+
+        # Compact flips
+        flip_to_tri = flip_to_tri[flip_to_tri >= 0]
+        flip_num = flip_to_tri.shape[0]
```

Comment:
This is the `compact_if_neg` section discussed in the benchmark results.

Reply:
I think the reason these lines are slow is essentially the same as the `unique`: both `unique` and advanced indexing seem to call slow `cupy.nonzero`. 😞